Test/Unit/Model/Api/ShippingMethodsTest.php

```diff
@@ -610,56 +722,439 @@ public function shippingEstimationWithDiscountAppliedToShipping()
-        $reflectionProperty->setValue($currentTestObject, $immutableQuoteMock);
+        $shippingAddress->expects($this->never())
+            ->method('setCollectShippingRates');
+
+        $addressRate = $this->getMockBuilder(\Magento\Quote\Model\Quote\Address\Rate::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $shippingRates = [['flatrate' => $addressRate]];
+        $shippingAddress->expects($this->never())
+            ->method('getGroupedAllShippingRates');
+
+        $quote = $this->getQuoteMock($shippingAddress);
+
+        $this->setUpRuleFactoryMock();
+        $this->initCurrentMock();
+
+        $this->cache->expects(self::once())->method('load')->with(self::anything())
+            ->willReturn(serialize($this->factoryShippingOptionsMock));
 
-        $testMethod = new \ReflectionMethod(BoltShippingMethods::class, 'couponInvalidForShippingAddress');
-        $testMethod->setAccessible(true);
+        $result = $this->currentMock->shippingEstimation($quote, $shippingAddressData);
 
-        $this->assertTrue($testMethod->invokeArgs($currentTestObject, [$parentQuoteCoupon]));
+        $this->assertEquals($this->factoryShippingOptionsMock, $result);
     }
 
     /**
-     * Get quote mock with quote items
-     *
-     * @param $shippingAddress
-     * @param $quoteId
-     * @param $parentQuoteId
-     * @return \PHPUnit_Framework_MockObject_MockObject
+     * @test
+     * @throws \ReflectionException
+     */
+    public function resetShippingCalculationIfNeeded()
+    {
+        $this->initCurrentMock();
+
+        $this->configHelper->expects(self::once())->method('getResetShippingCalculation')
+            ->with(self::STORE_ID)->willReturn(true);
+
+        $shippingAddressMock = $this->getMockBuilder(\Magento\Quote\Model\Quote\Address::class)
+            ->setMethods(['removeAllShippingRates', 'setCollectShippingRates'])
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $shippingAddressMock->expects(self::once())->method('removeAllShippingRates');
+        $shippingAddressMock->expects(self::once())->method('setCollectShippingRates')->with(true);
+
+        self::invokeInaccessibleMethod(
+            $this->currentMock,
+            'resetShippingCalculationIfNeeded',
+            [
+                $shippingAddressMock,
+                self::STORE_ID
+            ]
+        );
+    }
+
+    /**
+     * @test
+     */
+    public function shippingEstimation_freeShippingDiscount()
+    {
+        $email = "integration@bolt.com";
+        $shippingAddressData = [
+            'company'         => "",
+            'country'         => "United States",
+            'country_code'    => "US",
+            'email'           => $email,
+            'first_name'      => "John",
+            'last_name'       => "McCombs",
+            'locality'        => "Knoxville",
+            'phone'           => "+1 231 231 1234",
```

Comment:
The same as above regarding phone format.

Reply:
done